{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculate structural coverage and quality coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "file_record = \"./virus_records_ef_cf_plddt.csv\"\n",
    "df_record = pd.read_csv(file_record)\n",
    "file_protein = \"./virus_proteins_ef_cf_coverage.csv\"\n",
    "df_protein = pd.read_csv(file_protein)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>pept_cat</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>mat_pept_id</th>\n",
       "      <th>reg_id</th>\n",
       "      <th>Sort</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>protlen</th>\n",
       "      <th>translation</th>\n",
       "      <th>ef_pLDDT_mean</th>\n",
       "      <th>ef_pLDDT_0_50</th>\n",
       "      <th>ef_pLDDT_50_70</th>\n",
       "      <th>ef_pLDDT_70_90</th>\n",
       "      <th>ef_pLDDT_90_100</th>\n",
       "      <th>cf_pLDDT_mean</th>\n",
       "      <th>cf_pLDDT_0_50</th>\n",
       "      <th>cf_pLDDT_50_70</th>\n",
       "      <th>cf_pLDDT_70_90</th>\n",
       "      <th>cf_pLDDT_90_100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAA52564.1_4563</td>\n",
       "      <td>protein</td>\n",
       "      <td>CAA52564.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4563</td>\n",
       "      <td>[2713:&gt;2717](+)</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAA27332.1.11_7160</td>\n",
       "      <td>mat_pept</td>\n",
       "      <td>CAA27332.1</td>\n",
       "      <td>CAA27332.1.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7160</td>\n",
       "      <td>[2506:&gt;2507]</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAA47931.1.5_10363</td>\n",
       "      <td>mat_pept</td>\n",
       "      <td>AAA47931.1</td>\n",
       "      <td>AAA47931.1.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10363</td>\n",
       "      <td>[926:930]</td>\n",
       "      <td>4</td>\n",
       "      <td>NPAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAA27339.1.3_11752</td>\n",
       "      <td>mat_pept</td>\n",
       "      <td>CAA27339.1</td>\n",
       "      <td>CAA27339.1.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11752</td>\n",
       "      <td>[784:790]</td>\n",
       "      <td>6</td>\n",
       "      <td>TNRLTH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAI74981.1.3_11505</td>\n",
       "      <td>mat_pept</td>\n",
       "      <td>CAI74981.1</td>\n",
       "      <td>CAI74981.1.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11505</td>\n",
       "      <td>[492:499]</td>\n",
       "      <td>7</td>\n",
       "      <td>ASGRALA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               header  pept_cat  protein_id    mat_pept_id reg_id   Sort  \\\n",
       "0     CAA52564.1_4563   protein  CAA52564.1            NaN    NaN   4563   \n",
       "1  CAA27332.1.11_7160  mat_pept  CAA27332.1  CAA27332.1.11    NaN   7160   \n",
       "2  AAA47931.1.5_10363  mat_pept  AAA47931.1   AAA47931.1.5    NaN  10363   \n",
       "3  CAA27339.1.3_11752  mat_pept  CAA27339.1   CAA27339.1.3    NaN  11752   \n",
       "4  CAI74981.1.3_11505  mat_pept  CAI74981.1   CAI74981.1.3    NaN  11505   \n",
       "\n",
       "       coordinates  protlen translation  ef_pLDDT_mean  ef_pLDDT_0_50  \\\n",
       "0  [2713:>2717](+)        1           M            NaN            NaN   \n",
       "1     [2506:>2507]        1           G            NaN            NaN   \n",
       "2        [926:930]        4        NPAA            NaN            NaN   \n",
       "3        [784:790]        6      TNRLTH            NaN            NaN   \n",
       "4        [492:499]        7     ASGRALA            NaN            NaN   \n",
       "\n",
       "   ef_pLDDT_50_70  ef_pLDDT_70_90  ef_pLDDT_90_100  cf_pLDDT_mean  \\\n",
       "0             NaN             NaN              NaN            NaN   \n",
       "1             NaN             NaN              NaN            NaN   \n",
       "2             NaN             NaN              NaN            NaN   \n",
       "3             NaN             NaN              NaN            NaN   \n",
       "4             NaN             NaN              NaN            NaN   \n",
       "\n",
       "   cf_pLDDT_0_50  cf_pLDDT_50_70  cf_pLDDT_70_90  cf_pLDDT_90_100  \n",
       "0            NaN             NaN             NaN              NaN  \n",
       "1            NaN             NaN             NaN              NaN  \n",
       "2            NaN             NaN             NaN              NaN  \n",
       "3            NaN             NaN             NaN              NaN  \n",
       "4            NaN             NaN             NaN              NaN  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_record.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'YEHTATIPNVVGFPYKAHIERNXFSPMTLQLEVVXXSLEPTLNLEYITCEYKTVVPSPFIKCCGTSECSSKEQPDYQCKVYTGVYPFMWGGAYCFCDSENTQLSEAYVDRSDVCKHDHALAYKAHTASLKATIRISYGTINQTTEAFVNGEHAVNVGGSKFIFGPISTAWSPFDNKIVVYKDDVYNQDFPPYGSGQPGRFGDIQSRTVESKDLYANTALKLSRPSPGVVHVPYTQTPSGFKYWLKEKGSSLNTKAPFGCKIKTNPVRAMDCAVGSIPVSMDIPDSAFTRVVDAPAVTDLSCQVAVCTHSSDFGXVATLSYKTDKPGKCAVHSHSNVATLQEATVDVKEDGKVTVHFSXXSASPAFKVSVCDAKTTCTAACEPPKDHIVPYGASHNNQVFPDMSGTAMTWVQRMASGLGGLALIAVVVLVLVTCITMRR'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_record[df_record['header'] == 'AAA96330.1.5_6768']['translation'].values[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>Sort</th>\n",
       "      <th>pept_cat</th>\n",
       "      <th>protlen</th>\n",
       "      <th>ef_coverage</th>\n",
       "      <th>ef_plddt_0_50</th>\n",
       "      <th>ef_plddt_50_70</th>\n",
       "      <th>ef_plddt_70_90</th>\n",
       "      <th>ef_plddt_90_100</th>\n",
       "      <th>cf_coverage</th>\n",
       "      <th>cf_plddt_0_50</th>\n",
       "      <th>cf_plddt_50_70</th>\n",
       "      <th>cf_plddt_70_90</th>\n",
       "      <th>cf_plddt_90_100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAA52564.1_4563</td>\n",
       "      <td>CAA52564.1</td>\n",
       "      <td>4563</td>\n",
       "      <td>protein</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AIW68525.1_11911</td>\n",
       "      <td>AIW68525.1</td>\n",
       "      <td>11911</td>\n",
       "      <td>protein</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AIW68499.1_11911</td>\n",
       "      <td>AIW68499.1</td>\n",
       "      <td>11911</td>\n",
       "      <td>protein</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AOO54533.1_11976</td>\n",
       "      <td>AOO54533.1</td>\n",
       "      <td>11976</td>\n",
       "      <td>protein</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAA67112.1_4989</td>\n",
       "      <td>AAA67112.1</td>\n",
       "      <td>4989</td>\n",
       "      <td>protein</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71269</th>\n",
       "      <td>ASO76148.1_10247</td>\n",
       "      <td>ASO76148.1</td>\n",
       "      <td>10247</td>\n",
       "      <td>protein</td>\n",
       "      <td>7861</td>\n",
       "      <td>1488.0</td>\n",
       "      <td>388.0</td>\n",
       "      <td>299.0</td>\n",
       "      <td>660.0</td>\n",
       "      <td>141.0</td>\n",
       "      <td>1488.0</td>\n",
       "      <td>26.0</td>\n",
       "      <td>122.0</td>\n",
       "      <td>464.0</td>\n",
       "      <td>876.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71270</th>\n",
       "      <td>QRC47047.1_10244</td>\n",
       "      <td>QRC47047.1</td>\n",
       "      <td>10244</td>\n",
       "      <td>protein</td>\n",
       "      <td>8043</td>\n",
       "      <td>1357.0</td>\n",
       "      <td>192.0</td>\n",
       "      <td>305.0</td>\n",
       "      <td>693.0</td>\n",
       "      <td>167.0</td>\n",
       "      <td>1357.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>108.0</td>\n",
       "      <td>395.0</td>\n",
       "      <td>834.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71271</th>\n",
       "      <td>AIJ50565.1_10246</td>\n",
       "      <td>AIJ50565.1</td>\n",
       "      <td>10246</td>\n",
       "      <td>protein</td>\n",
       "      <td>8108</td>\n",
       "      <td>2219.0</td>\n",
       "      <td>519.0</td>\n",
       "      <td>756.0</td>\n",
       "      <td>761.0</td>\n",
       "      <td>183.0</td>\n",
       "      <td>2219.0</td>\n",
       "      <td>606.0</td>\n",
       "      <td>219.0</td>\n",
       "      <td>455.0</td>\n",
       "      <td>939.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71272</th>\n",
       "      <td>QRC47039.1_10254</td>\n",
       "      <td>QRC47039.1</td>\n",
       "      <td>10254</td>\n",
       "      <td>protein</td>\n",
       "      <td>8927</td>\n",
       "      <td>1392.0</td>\n",
       "      <td>173.0</td>\n",
       "      <td>184.0</td>\n",
       "      <td>686.0</td>\n",
       "      <td>349.0</td>\n",
       "      <td>1392.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>102.0</td>\n",
       "      <td>434.0</td>\n",
       "      <td>821.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71273</th>\n",
       "      <td>AYD75846.1_10227</td>\n",
       "      <td>AYD75846.1</td>\n",
       "      <td>10227</td>\n",
       "      <td>protein</td>\n",
       "      <td>13556</td>\n",
       "      <td>732.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>248.0</td>\n",
       "      <td>248.0</td>\n",
       "      <td>179.0</td>\n",
       "      <td>732.0</td>\n",
       "      <td>427.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>144.0</td>\n",
       "      <td>120.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>71274 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 header  protein_id   Sort pept_cat  protlen  ef_coverage  \\\n",
       "0       CAA52564.1_4563  CAA52564.1   4563  protein        1          0.0   \n",
       "1      AIW68525.1_11911  AIW68525.1  11911  protein        7          0.0   \n",
       "2      AIW68499.1_11911  AIW68499.1  11911  protein        8          0.0   \n",
       "3      AOO54533.1_11976  AOO54533.1  11976  protein        8          0.0   \n",
       "4       AAA67112.1_4989  AAA67112.1   4989  protein        9          0.0   \n",
       "...                 ...         ...    ...      ...      ...          ...   \n",
       "71269  ASO76148.1_10247  ASO76148.1  10247  protein     7861       1488.0   \n",
       "71270  QRC47047.1_10244  QRC47047.1  10244  protein     8043       1357.0   \n",
       "71271  AIJ50565.1_10246  AIJ50565.1  10246  protein     8108       2219.0   \n",
       "71272  QRC47039.1_10254  QRC47039.1  10254  protein     8927       1392.0   \n",
       "71273  AYD75846.1_10227  AYD75846.1  10227  protein    13556        732.0   \n",
       "\n",
       "       ef_plddt_0_50  ef_plddt_50_70  ef_plddt_70_90  ef_plddt_90_100  \\\n",
       "0                0.0             0.0             0.0              0.0   \n",
       "1                0.0             0.0             0.0              0.0   \n",
       "2                0.0             0.0             0.0              0.0   \n",
       "3                0.0             0.0             0.0              0.0   \n",
       "4                0.0             0.0             0.0              0.0   \n",
       "...              ...             ...             ...              ...   \n",
       "71269          388.0           299.0           660.0            141.0   \n",
       "71270          192.0           305.0           693.0            167.0   \n",
       "71271          519.0           756.0           761.0            183.0   \n",
       "71272          173.0           184.0           686.0            349.0   \n",
       "71273           57.0           248.0           248.0            179.0   \n",
       "\n",
       "       cf_coverage  cf_plddt_0_50  cf_plddt_50_70  cf_plddt_70_90  \\\n",
       "0              0.0            0.0             0.0             0.0   \n",
       "1              0.0            0.0             0.0             0.0   \n",
       "2              0.0            0.0             0.0             0.0   \n",
       "3              0.0            0.0             0.0             0.0   \n",
       "4              0.0            0.0             0.0             0.0   \n",
       "...            ...            ...             ...             ...   \n",
       "71269       1488.0           26.0           122.0           464.0   \n",
       "71270       1357.0           20.0           108.0           395.0   \n",
       "71271       2219.0          606.0           219.0           455.0   \n",
       "71272       1392.0           35.0           102.0           434.0   \n",
       "71273        732.0          427.0            41.0           144.0   \n",
       "\n",
       "       cf_plddt_90_100  \n",
       "0                  0.0  \n",
       "1                  0.0  \n",
       "2                  0.0  \n",
       "3                  0.0  \n",
       "4                  0.0  \n",
       "...                ...  \n",
       "71269            876.0  \n",
       "71270            834.0  \n",
       "71271            939.0  \n",
       "71272            821.0  \n",
       "71273            120.0  \n",
       "\n",
       "[71274 rows x 15 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "ESMFold\n",
    "'''\n",
    "\n",
    "n = 0\n",
    "for index, row in df_protein.iterrows():\n",
    "\n",
    "    n += 1\n",
    "    header = row['header']\n",
    "    protein_id = row['protein_id']\n",
    "\n",
    "    df_record_sub = df_record[(df_record['protein_id'] == protein_id) &\n",
    "                              (df_record['ef_pLDDT_mean'].notnull())]\n",
    "    \n",
    "    # we didn't predict any structures for this protein\n",
    "    if df_record_sub.shape[0] == 0:\n",
    "        ef_coverage = 0\n",
    "        ef_plddt_0_50 = 0\n",
    "        ef_plddt_50_70 = 0\n",
    "        ef_plddt_70_90 = 0\n",
    "        ef_plddt_90_100 = 0\n",
    "    # we predicted structure for this protein\n",
    "    elif header in df_record_sub['header'].values:\n",
    "        ef_coverage = df_record_sub[df_record_sub['header'] == header]['protlen'].values[0]\n",
    "        ef_plddt_0_50 = df_record_sub[df_record_sub['header'] == header]['ef_pLDDT_0_50'].values[0]\n",
    "        ef_plddt_50_70 = df_record_sub[df_record_sub['header'] == header]['ef_pLDDT_50_70'].values[0]\n",
    "        ef_plddt_70_90 = df_record_sub[df_record_sub['header'] == header]['ef_pLDDT_70_90'].values[0]\n",
    "        ef_plddt_90_100 = df_record_sub[df_record_sub['header'] == header]['ef_pLDDT_90_100'].values[0]\n",
    "    # we predicted structures for peptides or regions of this protein\n",
    "    else:\n",
    "        ef_coverage = np.nan\n",
    "        ef_plddt_0_50 = np.nan\n",
    "        ef_plddt_50_70 = np.nan\n",
    "        ef_plddt_70_90 = np.nan\n",
    "        ef_plddt_90_100 = np.nan\n",
    "\n",
    "    df_protein.at[index, 'ef_coverage'] = ef_coverage\n",
    "    df_protein.at[index, 'ef_plddt_0_50'] = ef_plddt_0_50\n",
    "    df_protein.at[index, 'ef_plddt_50_70'] = ef_plddt_50_70\n",
    "    df_protein.at[index, 'ef_plddt_70_90'] = ef_plddt_70_90\n",
    "    df_protein.at[index, 'ef_plddt_90_100'] = ef_plddt_90_100\n",
    "\n",
    "    print(n, header, ef_coverage, ef_plddt_0_50, ef_plddt_50_70, ef_plddt_70_90, ef_plddt_90_100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "ColabFold\n",
    "'''\n",
    "\n",
    "n = 0\n",
    "for index, row in df_protein.iterrows():\n",
    "\n",
    "    n += 1\n",
    "    header = row['header']\n",
    "    protein_id = row['protein_id']\n",
    "\n",
    "    df_record_sub = df_record[(df_record['protein_id'] == protein_id) &\n",
    "                              (df_record['cf_pLDDT_mean'].notnull())]\n",
    "    \n",
    "    # we didn't predict any structures for this protein\n",
    "    if df_record_sub.shape[0] == 0:\n",
    "        ef_coverage = 0\n",
    "        ef_plddt_0_50 = 0\n",
    "        ef_plddt_50_70 = 0\n",
    "        ef_plddt_70_90 = 0\n",
    "        ef_plddt_90_100 = 0\n",
    "    # we predicted structure for this protein\n",
    "    elif header in df_record_sub['header'].values:\n",
    "        ef_coverage = df_record_sub[df_record_sub['header'] == header]['protlen'].values[0]\n",
    "        ef_plddt_0_50 = df_record_sub[df_record_sub['header'] == header]['cf_pLDDT_0_50'].values[0]\n",
    "        ef_plddt_50_70 = df_record_sub[df_record_sub['header'] == header]['cf_pLDDT_50_70'].values[0]\n",
    "        ef_plddt_70_90 = df_record_sub[df_record_sub['header'] == header]['cf_pLDDT_70_90'].values[0]\n",
    "        ef_plddt_90_100 = df_record_sub[df_record_sub['header'] == header]['cf_pLDDT_90_100'].values[0]\n",
    "    # we predicted structures for peptides or regions of this protein\n",
    "    else:\n",
    "        ef_coverage = np.nan\n",
    "        ef_plddt_0_50 = np.nan\n",
    "        ef_plddt_50_70 = np.nan\n",
    "        ef_plddt_70_90 = np.nan\n",
    "        ef_plddt_90_100 = np.nan\n",
    "\n",
    "    df_protein.at[index, 'cf_coverage'] = ef_coverage\n",
    "    df_protein.at[index, 'cf_plddt_0_50'] = ef_plddt_0_50\n",
    "    df_protein.at[index, 'cf_plddt_50_70'] = ef_plddt_50_70\n",
    "    df_protein.at[index, 'cf_plddt_70_90'] = ef_plddt_70_90\n",
    "    df_protein.at[index, 'cf_plddt_90_100'] = ef_plddt_90_100\n",
    "\n",
    "    print(n, header, ef_coverage, ef_plddt_0_50, ef_plddt_50_70, ef_plddt_70_90, ef_plddt_90_100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>Sort</th>\n",
       "      <th>pept_cat</th>\n",
       "      <th>protlen</th>\n",
       "      <th>ef_coverage</th>\n",
       "      <th>ef_plddt_0_50</th>\n",
       "      <th>ef_plddt_50_70</th>\n",
       "      <th>ef_plddt_70_90</th>\n",
       "      <th>ef_plddt_90_100</th>\n",
       "      <th>cf_coverage</th>\n",
       "      <th>cf_plddt_0_50</th>\n",
       "      <th>cf_plddt_50_70</th>\n",
       "      <th>cf_plddt_70_90</th>\n",
       "      <th>cf_plddt_90_100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [header, protein_id, Sort, pept_cat, protlen, ef_coverage, ef_plddt_0_50, ef_plddt_50_70, ef_plddt_70_90, ef_plddt_90_100, cf_coverage, cf_plddt_0_50, cf_plddt_50_70, cf_plddt_70_90, cf_plddt_90_100]\n",
       "Index: []"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_protein[df_protein['cf_coverage'].isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_protein.to_csv(\"virus_proteins_ef_cf_coverage.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statistics\n",
    "\n",
    "def plddt_mean(plddt: list, round_to: int = 1):\n",
    "    '''\n",
    "    Calculate the average pLDDT score.\n",
    "    '''\n",
    "    return round(statistics.mean(plddt), round_to)\n",
    "\n",
    "def plddt_coverage_aa_number(plddt: list, lower_threshold: float,\n",
    "                             upper_threshold: float, include_upper_threshold: bool = False):\n",
    "    '''\n",
    "    Calculate the coverage of a protein sequence in number of AA\n",
    "    based on a pLDDT thresholds and return the number of AA.\n",
    "    '''\n",
    "    if include_upper_threshold:\n",
    "        return len([i for i in plddt if i >= lower_threshold and i <= upper_threshold])\n",
    "    else:\n",
    "        return len([i for i in plddt if i >= lower_threshold and i < upper_threshold])\n",
    "\n",
    "def plddt_coverage_across_default_categories(plddt: list):\n",
    "    '''\n",
    "    Calculate the coverage of a protein sequence in number of AA\n",
    "    based on default pLDDT categories: [0, 50), [50, 70), [70, 90), [90, 100] -\n",
    "    and return a list with the number of AA for each category.\n",
    "    '''\n",
    "    coverage = []\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 0, 50))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 50, 70))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 70, 90))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 90, 100, include_upper_threshold=True))\n",
    "    return coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# open jsof file with plddt scores\n",
    "file_ef_plddt = \"../plddt_scores/all_ESMFold_plddt_scores.json\"\n",
    "\n",
    "with open(file_ef_plddt, 'r') as f:\n",
    "    ef_plddt = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "ESMFold\n",
    "'''\n",
    "\n",
    "combined_plddt = {}\n",
    "n = 0\n",
    "\n",
    "for index, row in df_protein[df_protein['ef_coverage'].isnull()].iterrows():\n",
    "    \n",
    "    n += 1\n",
    "    header = row['header']\n",
    "    protein_id = row['protein_id']\n",
    "    protlen = row['protlen']\n",
    "\n",
    "    combined_plddt[header] = {}\n",
    "\n",
    "    df_record_sub = df_record[(df_record['protein_id'] == protein_id) &\n",
    "                              (df_record['ef_pLDDT_mean'].notnull())].sort_values(by='protlen', ascending=False)\n",
    "    \n",
    "    record_id_list = [np.nan] * protlen\n",
    "    plddt_list = [np.nan] * protlen\n",
    "    \n",
    "    for index2, row2 in df_record_sub.iterrows():\n",
    "        # coordinates are in Spyros' format:\n",
    "        # first position is not part of the protein, while the last one is\n",
    "        start, end = row2['coordinates'].lstrip('[').rstrip(']').split(':')\n",
    "        start = int(start.lstrip('<'))\n",
    "        end = int(end.lstrip('>'))\n",
    "        \n",
    "        record_id = row2['header']\n",
    "        record_plddt = ef_plddt[record_id]\n",
    "\n",
    "        for i in range(start, end):\n",
    "            if record_id_list[i] is np.nan:\n",
    "                try:\n",
    "                    record_id_list[i] = record_id\n",
    "                    plddt_list[i] = record_plddt[i - start]\n",
    "                except IndexError:\n",
    "                    print('ERROR: ', record_id, end - start, len(record_plddt))\n",
    "                    \n",
    "    combined_plddt[header]['record_id'] = record_id_list\n",
    "    combined_plddt[header]['plddt'] = plddt_list\n",
    "\n",
    "    ef_coverage = len([i for i in plddt_list if i is not np.nan])\n",
    "    ef_plddt_0_50, ef_plddt_50_70, ef_plddt_70_90, ef_plddt_90_100 = plddt_coverage_across_default_categories(plddt_list)\n",
    "\n",
    "    df_protein.at[index, 'ef_coverage'] = ef_coverage\n",
    "    df_protein.at[index, 'ef_plddt_0_50'] = ef_plddt_0_50\n",
    "    df_protein.at[index, 'ef_plddt_50_70'] = ef_plddt_50_70\n",
    "    df_protein.at[index, 'ef_plddt_70_90'] = ef_plddt_70_90\n",
    "    df_protein.at[index, 'ef_plddt_90_100'] = ef_plddt_90_100\n",
    "\n",
    "    print(n, header, ef_coverage, ef_plddt_0_50, ef_plddt_50_70, ef_plddt_70_90, ef_plddt_90_100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the combined pLDDT scores as a json file\n",
    "with open('ef_plddt_coverage_polyproteins_from_peptides_and_regions.json', 'w') as f:\n",
    "    json.dump(combined_plddt, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# open jsof file with plddt scores\n",
    "file_ef_plddt = \"../plddt_scores/all_rank_001_CF_plddt_scores.json\"\n",
    "\n",
    "with open(file_ef_plddt, 'r') as f:\n",
    "    ef_plddt = json.load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "ColabFold - pLDDT scores\n",
    "'''\n",
    "\n",
    "combined_plddt = {}\n",
    "n = 0\n",
    "\n",
    "for index, row in df_protein[df_protein['cf_coverage'].isnull()].iterrows():\n",
    "    \n",
    "    n += 1\n",
    "    header = row['header']\n",
    "    protein_id = row['protein_id']\n",
    "    protlen = row['protlen']\n",
    "\n",
    "    combined_plddt[header] = {}\n",
    "\n",
    "    df_record_sub = df_record[(df_record['protein_id'] == protein_id) &\n",
    "                              (df_record['cf_pLDDT_mean'].notnull())].sort_values(by='protlen', ascending=False)\n",
    "    \n",
    "    record_id_list = [np.nan] * protlen\n",
    "    plddt_list = [np.nan] * protlen\n",
    "    \n",
    "    for index2, row2 in df_record_sub.iterrows():\n",
    "        # coordinates are in Spyros' format:\n",
    "        # first position is not part of the protein, while the last one is\n",
    "        start, end = row2['coordinates'].lstrip('[').rstrip(']').split(':')\n",
    "        start = int(start.lstrip('<'))\n",
    "        end = int(end.lstrip('>'))\n",
    "        \n",
    "        record_id = row2['header']\n",
    "        record_plddt = ef_plddt[record_id]\n",
    "\n",
    "        for i in range(start, end):\n",
    "            if record_id_list[i] is np.nan:\n",
    "                try:\n",
    "                    record_id_list[i] = record_id\n",
    "                    plddt_list[i] = record_plddt[i - start]\n",
    "                except IndexError:\n",
    "                    print('ERROR: ', record_id, end - start, len(record_plddt))\n",
    "                    \n",
    "    combined_plddt[header]['record_id'] = record_id_list\n",
    "    combined_plddt[header]['plddt'] = plddt_list\n",
    "\n",
    "    ef_coverage = len([i for i in plddt_list if i is not np.nan])\n",
    "    ef_plddt_0_50, ef_plddt_50_70, ef_plddt_70_90, ef_plddt_90_100 = plddt_coverage_across_default_categories(plddt_list)\n",
    "\n",
    "    df_protein.at[index, 'cf_coverage'] = ef_coverage\n",
    "    df_protein.at[index, 'cf_plddt_0_50'] = ef_plddt_0_50\n",
    "    df_protein.at[index, 'cf_plddt_50_70'] = ef_plddt_50_70\n",
    "    df_protein.at[index, 'cf_plddt_70_90'] = ef_plddt_70_90\n",
    "    df_protein.at[index, 'cf_plddt_90_100'] = ef_plddt_90_100\n",
    "\n",
    "    print(n, header, ef_coverage, ef_plddt_0_50, ef_plddt_50_70, ef_plddt_70_90, ef_plddt_90_100)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the combined pLDDT scores as a json file\n",
    "with open('cf_plddt_coverage_polyproteins_from_peptides_and_regions.json', 'w') as f:\n",
    "    json.dump(combined_plddt, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Structural coverage"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Calculate the per residue coverage of the viral protein\n",
    "by ESMFold predictions\n",
    "'''\n",
    "\n",
    "n = 0\n",
    "\n",
    "for index, row in df[df['pept_cat'] == 'protein'].iterrows():\n",
    "    \n",
    "    n += 1\n",
    "\n",
    "    protein_id = row['protein_id']\n",
    "    protlen = row['protlen']\n",
    "\n",
    "    # get the sub dataframe for the peptides and regions of the protein that are predicted by ESMFold\n",
    "    df_sub = df[(df['protein_id'] == protein_id) & (df['ef_pLDDT_mean'].notnull())]\n",
    "\n",
    "    # make a list of 0s the length of the protein\n",
    "    coverage_list = [0] * protlen\n",
    "\n",
    "    # for each predicted peptide or region, add 1 to the list of 0s for each amino acid in the coordinates\n",
    "    for index2, row2 in df_sub.iterrows():\n",
    "\n",
    "        if row2['pept_cat'] == 'protein':\n",
    "            coverage_list = [i + 1 for i in coverage_list]\n",
    "        else:\n",
    "            # coordinates are in the Spyros' format (first position is not part of the protein, while the last one is) \n",
    "            start, end = row2['coordinates'].lstrip('[').rstrip(']').split(':')\n",
    "            start = int(start.lstrip('<'))\n",
    "            end = int(end.lstrip('>'))\n",
    "            \n",
    "            for i in range(int(start), int(end)):\n",
    "                coverage_list[i] += 1\n",
    "            \n",
    "    coverage = coverage_list.count(1) + coverage_list.count(2)\n",
    "    double_coverage = coverage_list.count(2)\n",
    "    df.at[index, 'ef_coverage'] = coverage\n",
    "    df.at[index, 'ef_double_coverage'] = double_coverage\n",
    "\n",
    "    print(n, protein_id, round(coverage/ protlen, 2))\n",
    "    #print(n, protein_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Calculate the per residue coverage of the viral protein\n",
    "by ColabFold predictions\n",
    "'''\n",
    "\n",
    "n = 0\n",
    "\n",
    "for index, row in df[df['pept_cat'] == 'protein'].iterrows():\n",
    "    \n",
    "    n += 1\n",
    "\n",
    "    header = row['header']\n",
    "    protein_id = row['protein_id']\n",
    "    protlen = row['protlen']\n",
    "\n",
    "    # get the sub dataframe for the peptides and regions of the protein that are predicted by ESMFold\n",
    "    df_sub = df[(df['protein_id'] == protein_id) & (df['cf_pLDDT_mean'].notnull())]\n",
    "\n",
    "    # make a list of 0s the length of the protein\n",
    "    coverage_list = [0] * protlen\n",
    "\n",
    "    # for each predicted peptide or region, add 1 to the list of 0s for each amino acid in the coordinates\n",
    "    for index2, row2 in df_sub.iterrows():\n",
    "\n",
    "        if row2['protlen'] == protlen:\n",
    "            coverage_list = [i + 1 for i in coverage_list]\n",
    "        else:\n",
    "            # coordinates are in the Spyros' format (first position is not part of the protein, while the last one is) \n",
    "            start, end = row2['coordinates'].lstrip('[').rstrip(']').split(':')\n",
    "            start = int(start.lstrip('<'))\n",
    "            end = int(end.lstrip('>'))\n",
    "            \n",
    "            for i in range(int(start), int(end)):\n",
    "                coverage_list[i] += 1\n",
    "            \n",
    "    coverage = coverage_list.count(1) + coverage_list.count(2)\n",
    "    double_coverage = coverage_list.count(2)\n",
    "    df.at[index, 'cf_coverage'] = coverage\n",
    "    df.at[index, 'cf_double_coverage'] = double_coverage\n",
    "\n",
    "    print(n, protein_id, round(coverage/ protlen, 2))\n",
    "    #print(n, protein_id)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "n = 0\n",
    "\n",
    "for index, row in df[df['pept_cat'] == 'protein'].iterrows():\n",
    "    \n",
    "    n += 1\n",
    "    protein_id = row['protein_id']\n",
    "    pept_cat = row['pept_cat']\n",
    "    protlen = row['protlen']\n",
    "    prot_seq = row['translation']\n",
    "\n",
    "    # check that ef_pLDDT_mean is not NaN\n",
    "    if not np.isnan(row['ef_pLDDT_mean']):\n",
    "        # add protlen to the 'ef_coverage' column\n",
    "        ef_coverage = protlen\n",
    "        df.at[index, 'ef_coverage'] = ef_coverage\n",
    "    else:\n",
    "        # get the sub dataframe for the peptides and regions of the protein that are predicted by ESMFold\n",
    "        df_sub = df[(df['protein_id'] == protein_id) & (df['pept_cat'] != 'protein') & (df['ef_pLDDT_mean'].notnull())]\n",
    "\n",
    "        # make a list of 0s the length of the protein\n",
    "        ef_coverage_list = [0] * protlen\n",
    "\n",
    "        # for each predicted peptide or region, add 1 to the list of 0s for each amino acid in the coordinates\n",
    "        for index2, row2 in df_sub.iterrows():\n",
    "            \n",
    "            # coordinates are in the Spyros' format (first position is not part of the protein, while the last one is) \n",
    "            start, end = row2['coordinates'].lstrip('[').rstrip(']').split(':')\n",
    "            \n",
    "            start = int(start.lstrip('<'))\n",
    "            end = int(end.lstrip('>'))\n",
    "            \n",
    "            for i in range(int(start), int(end)):\n",
    "                ef_coverage_list[i] += 1\n",
    "            \n",
    "            ef_coverage = ef_coverage_list.count(1)\n",
    "            ef_double_coverage = ef_coverage_list.count(2)\n",
    "            df.at[index, 'ef_coverage'] = ef_coverage\n",
    "            df.at[index, 'ef_double_coverage'] = ef_double_coverage\n",
    "\n",
    "    #print(n, protein_id, round(ef_coverage/ protlen, 2))\n",
    "    print(n, protein_id)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# pLDDT coverage across default categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Calculate the average pLDDT score and pLDDT coverage across default categories for all ESMFold/ ColabFold models.\n",
    "'''\n",
    "\n",
    "import pandas as pd\n",
    "import json\n",
    "import statistics\n",
    "\n",
    "def plddt_mean(plddt: list, round_to: int = 1):\n",
    "    '''\n",
    "    Calculate the average pLDDT score.\n",
    "    '''\n",
    "    return round(statistics.mean(plddt), round_to)\n",
    "\n",
    "def plddt_coverage_aa_number(plddt: list, lower_threshold: float,\n",
    "                             upper_threshold: float, include_upper_threshold: bool = False):\n",
    "    '''\n",
    "    Calculate the coverage of a protein sequence in number of AA\n",
    "    based on a pLDDT thresholds and return the number of AA.\n",
    "    '''\n",
    "    if include_upper_threshold:\n",
    "        return len([i for i in plddt if i >= lower_threshold and i <= upper_threshold])\n",
    "    else:\n",
    "        return len([i for i in plddt if i >= lower_threshold and i < upper_threshold])\n",
    "\n",
    "def plddt_coverage_across_default_categories(plddt: list):\n",
    "    '''\n",
    "    Calculate the coverage of a protein sequence in number of AA\n",
    "    based on default pLDDT categories: [0, 50), [50, 70), [70, 90), [90, 100] -\n",
    "    and return a list with the number of AA for each category.\n",
    "    '''\n",
    "    coverage = []\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 0, 50))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 50, 70))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 70, 90))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 90, 100, include_upper_threshold=True))\n",
    "    return coverage\n",
    "\n",
    "file_meta = \"../virus_proteomes/virus_records.csv\"\n",
    "df_meta = pd.read_csv(file_meta)\n",
    "\n",
    "ef_json_file = \"../virus_proteomes/all_ESMFold_plddt_scores.json\"\n",
    "with open(ef_json_file, 'r') as f:\n",
    "    ef_plddt = json.load(f)\n",
    "\n",
    "n = 0\n",
    "print(\"index\\trecord_id\\tef_plddt_mean\\tef_plddt_0_50\\tef_plddt_50_70\\tef_plddt_70_90\\tef_plddt_90_100\")\n",
    "for record_id, plddt in ef_plddt.items():\n",
    "    n += 1\n",
    "    plddt_cov = plddt_coverage_across_default_categories(plddt)\n",
    "    plddt_mean_value = plddt_mean(plddt)\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'ef_pLDDT_mean'] = plddt_mean_value\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'ef_pLDDT_0_50'] = plddt_cov[0]\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'ef_pLDDT_50_70'] = plddt_cov[1]\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'ef_pLDDT_70_90'] = plddt_cov[2]\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'ef_pLDDT_90_100'] = plddt_cov[3]\n",
    "    print(n, record_id, plddt_mean_value, plddt_cov[0], plddt_cov[1], plddt_cov[2], plddt_cov[3], sep='\\t')\n",
    "\n",
    "df_meta.to_csv(\"./virus_records_ef_plddt.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cf_file = \"../plddt_scores/all_rank_001_CF_plddt_scores.json\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "Calculate the average pLDDT score and pLDDT coverage across default categories for all ESMFold/ ColabFold models.\n",
    "'''\n",
    "\n",
    "import pandas as pd\n",
    "import json\n",
    "import statistics\n",
    "\n",
    "def plddt_mean(plddt: list, round_to: int = 1):\n",
    "    '''\n",
    "    Calculate the average pLDDT score.\n",
    "    '''\n",
    "    return round(statistics.mean(plddt), round_to)\n",
    "\n",
    "def plddt_coverage_aa_number(plddt: list, lower_threshold: float,\n",
    "                             upper_threshold: float, include_upper_threshold: bool = False):\n",
    "    '''\n",
    "    Calculate the coverage of a protein sequence in number of AA\n",
    "    based on a pLDDT thresholds and return the number of AA.\n",
    "    '''\n",
    "    if include_upper_threshold:\n",
    "        return len([i for i in plddt if i >= lower_threshold and i <= upper_threshold])\n",
    "    else:\n",
    "        return len([i for i in plddt if i >= lower_threshold and i < upper_threshold])\n",
    "\n",
    "def plddt_coverage_across_default_categories(plddt: list):\n",
    "    '''\n",
    "    Calculate the coverage of a protein sequence in number of AA\n",
    "    based on default pLDDT categories: [0, 50), [50, 70), [70, 90), [90, 100] -\n",
    "    and return a list with the number of AA for each category.\n",
    "    '''\n",
    "    coverage = []\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 0, 50))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 50, 70))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 70, 90))\n",
    "    coverage.append(plddt_coverage_aa_number(plddt, 90, 100, include_upper_threshold=True))\n",
    "    return coverage\n",
    "\n",
    "file_meta = \"./virus_records_ef_plddt.csv\"\n",
    "df_meta = pd.read_csv(file_meta)\n",
    "\n",
    "cf_json_file = \"../plddt_scores/all_rank_001_CF_plddt_scores.json\"\n",
    "with open(cf_json_file, 'r') as f:\n",
    "    cf_plddt = json.load(f)\n",
    "\n",
    "n = 0\n",
    "print(\"index\\trecord_id\\tef_plddt_mean\\tef_plddt_0_50\\tef_plddt_50_70\\tef_plddt_70_90\\tef_plddt_90_100\")\n",
    "for record_id, plddt in cf_plddt.items():\n",
    "    n += 1\n",
    "    plddt_cov = plddt_coverage_across_default_categories(plddt)\n",
    "    plddt_mean_value = plddt_mean(plddt)\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'cf_pLDDT_mean'] = plddt_mean_value\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'cf_pLDDT_0_50'] = plddt_cov[0]\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'cf_pLDDT_50_70'] = plddt_cov[1]\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'cf_pLDDT_70_90'] = plddt_cov[2]\n",
    "    df_meta.loc[df_meta['header'] == record_id, 'cf_pLDDT_90_100'] = plddt_cov[3]\n",
    "    print(n, record_id, plddt_mean_value, plddt_cov[0], plddt_cov[1], plddt_cov[2], plddt_cov[3], sep='\\t')\n",
    "\n",
    "df_meta.to_csv(\"./virus_records_ef_cf_plddt.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>header</th>\n",
       "      <th>pept_cat</th>\n",
       "      <th>protein_id</th>\n",
       "      <th>mat_pept_id</th>\n",
       "      <th>reg_id</th>\n",
       "      <th>Sort</th>\n",
       "      <th>coordinates</th>\n",
       "      <th>protlen</th>\n",
       "      <th>translation</th>\n",
       "      <th>ef_pLDDT_mean</th>\n",
       "      <th>ef_pLDDT_0_50</th>\n",
       "      <th>ef_pLDDT_50_70</th>\n",
       "      <th>ef_pLDDT_70_90</th>\n",
       "      <th>ef_pLDDT_90_100</th>\n",
       "      <th>cf_pLDDT_mean</th>\n",
       "      <th>cf_pLDDT_0_50</th>\n",
       "      <th>cf_pLDDT_50_70</th>\n",
       "      <th>cf_pLDDT_70_90</th>\n",
       "      <th>cf_pLDDT_90_100</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CAA52564.1_4563</td>\n",
       "      <td>protein</td>\n",
       "      <td>CAA52564.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4563</td>\n",
       "      <td>[2713:&gt;2717](+)</td>\n",
       "      <td>1</td>\n",
       "      <td>M</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CAA27332.1.11_7160</td>\n",
       "      <td>mat_pept</td>\n",
       "      <td>CAA27332.1</td>\n",
       "      <td>CAA27332.1.11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7160</td>\n",
       "      <td>[2506:&gt;2507]</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAA47931.1.5_10363</td>\n",
       "      <td>mat_pept</td>\n",
       "      <td>AAA47931.1</td>\n",
       "      <td>AAA47931.1.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10363</td>\n",
       "      <td>[926:930]</td>\n",
       "      <td>4</td>\n",
       "      <td>NPAA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CAA27339.1.3_11752</td>\n",
       "      <td>mat_pept</td>\n",
       "      <td>CAA27339.1</td>\n",
       "      <td>CAA27339.1.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11752</td>\n",
       "      <td>[784:790]</td>\n",
       "      <td>6</td>\n",
       "      <td>TNRLTH</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CAI74981.1.3_11505</td>\n",
       "      <td>mat_pept</td>\n",
       "      <td>CAI74981.1</td>\n",
       "      <td>CAI74981.1.3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11505</td>\n",
       "      <td>[492:499]</td>\n",
       "      <td>7</td>\n",
       "      <td>ASGRALA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87135</th>\n",
       "      <td>ASO76148.1_10247</td>\n",
       "      <td>protein</td>\n",
       "      <td>ASO76148.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10247</td>\n",
       "      <td>join{[855:17505](+), [17504:24440](+)}</td>\n",
       "      <td>7861</td>\n",
       "      <td>MVLIVPLSRTNRPPITKPRHFHALPNYYTKYGSDIIKRNIYYNSVD...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87136</th>\n",
       "      <td>QRC47047.1_10244</td>\n",
       "      <td>protein</td>\n",
       "      <td>QRC47047.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10244</td>\n",
       "      <td>join{[409:17569](+), [17568:24540](+)}</td>\n",
       "      <td>8043</td>\n",
       "      <td>MVFIHPNFVYLSKGLGKPRCTASTIVQREYQYQYISYPHKTVGSIF...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87137</th>\n",
       "      <td>AIJ50565.1_10246</td>\n",
       "      <td>protein</td>\n",
       "      <td>AIJ50565.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10246</td>\n",
       "      <td>join{[1017:18411](+), [18410:25343](+)}</td>\n",
       "      <td>8108</td>\n",
       "      <td>MVLIVPLAYTNRPPICKPRRYSVLPSYHHKYGSDMIKRNIYYRGTD...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87138</th>\n",
       "      <td>QRC47039.1_10254</td>\n",
       "      <td>protein</td>\n",
       "      <td>QRC47039.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10254</td>\n",
       "      <td>join{[399:20106](+), [20105:27182](+)}</td>\n",
       "      <td>8927</td>\n",
       "      <td>MVVLQVVAQNATKSGETPGWRLPLKSLEKGKAIARKSMGSSLTLTQ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87139</th>\n",
       "      <td>AYD75846.1_10227</td>\n",
       "      <td>protein</td>\n",
       "      <td>AYD75846.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10227</td>\n",
       "      <td>[128:40799](+)</td>\n",
       "      <td>13556</td>\n",
       "      <td>MCDHTVYQRDAINEQNLIRKDGVQKQTYVIIYDDDDVQNGSRLNED...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>87140 rows × 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   header  pept_cat  protein_id    mat_pept_id reg_id   Sort  \\\n",
       "0         CAA52564.1_4563   protein  CAA52564.1            NaN    NaN   4563   \n",
       "1      CAA27332.1.11_7160  mat_pept  CAA27332.1  CAA27332.1.11    NaN   7160   \n",
       "2      AAA47931.1.5_10363  mat_pept  AAA47931.1   AAA47931.1.5    NaN  10363   \n",
       "3      CAA27339.1.3_11752  mat_pept  CAA27339.1   CAA27339.1.3    NaN  11752   \n",
       "4      CAI74981.1.3_11505  mat_pept  CAI74981.1   CAI74981.1.3    NaN  11505   \n",
       "...                   ...       ...         ...            ...    ...    ...   \n",
       "87135    ASO76148.1_10247   protein  ASO76148.1            NaN    NaN  10247   \n",
       "87136    QRC47047.1_10244   protein  QRC47047.1            NaN    NaN  10244   \n",
       "87137    AIJ50565.1_10246   protein  AIJ50565.1            NaN    NaN  10246   \n",
       "87138    QRC47039.1_10254   protein  QRC47039.1            NaN    NaN  10254   \n",
       "87139    AYD75846.1_10227   protein  AYD75846.1            NaN    NaN  10227   \n",
       "\n",
       "                                   coordinates  protlen  \\\n",
       "0                              [2713:>2717](+)        1   \n",
       "1                                 [2506:>2507]        1   \n",
       "2                                    [926:930]        4   \n",
       "3                                    [784:790]        6   \n",
       "4                                    [492:499]        7   \n",
       "...                                        ...      ...   \n",
       "87135   join{[855:17505](+), [17504:24440](+)}     7861   \n",
       "87136   join{[409:17569](+), [17568:24540](+)}     8043   \n",
       "87137  join{[1017:18411](+), [18410:25343](+)}     8108   \n",
       "87138   join{[399:20106](+), [20105:27182](+)}     8927   \n",
       "87139                           [128:40799](+)    13556   \n",
       "\n",
       "                                             translation  ef_pLDDT_mean  \\\n",
       "0                                                      M            NaN   \n",
       "1                                                      G            NaN   \n",
       "2                                                   NPAA            NaN   \n",
       "3                                                 TNRLTH            NaN   \n",
       "4                                                ASGRALA            NaN   \n",
       "...                                                  ...            ...   \n",
       "87135  MVLIVPLSRTNRPPITKPRHFHALPNYYTKYGSDIIKRNIYYNSVD...            NaN   \n",
       "87136  MVFIHPNFVYLSKGLGKPRCTASTIVQREYQYQYISYPHKTVGSIF...            NaN   \n",
       "87137  MVLIVPLAYTNRPPICKPRRYSVLPSYHHKYGSDMIKRNIYYRGTD...            NaN   \n",
       "87138  MVVLQVVAQNATKSGETPGWRLPLKSLEKGKAIARKSMGSSLTLTQ...            NaN   \n",
       "87139  MCDHTVYQRDAINEQNLIRKDGVQKQTYVIIYDDDDVQNGSRLNED...            NaN   \n",
       "\n",
       "       ef_pLDDT_0_50  ef_pLDDT_50_70  ef_pLDDT_70_90  ef_pLDDT_90_100  \\\n",
       "0                NaN             NaN             NaN              NaN   \n",
       "1                NaN             NaN             NaN              NaN   \n",
       "2                NaN             NaN             NaN              NaN   \n",
       "3                NaN             NaN             NaN              NaN   \n",
       "4                NaN             NaN             NaN              NaN   \n",
       "...              ...             ...             ...              ...   \n",
       "87135            NaN             NaN             NaN              NaN   \n",
       "87136            NaN             NaN             NaN              NaN   \n",
       "87137            NaN             NaN             NaN              NaN   \n",
       "87138            NaN             NaN             NaN              NaN   \n",
       "87139            NaN             NaN             NaN              NaN   \n",
       "\n",
       "       cf_pLDDT_mean  cf_pLDDT_0_50  cf_pLDDT_50_70  cf_pLDDT_70_90  \\\n",
       "0                NaN            NaN             NaN             NaN   \n",
       "1                NaN            NaN             NaN             NaN   \n",
       "2                NaN            NaN             NaN             NaN   \n",
       "3                NaN            NaN             NaN             NaN   \n",
       "4                NaN            NaN             NaN             NaN   \n",
       "...              ...            ...             ...             ...   \n",
       "87135            NaN            NaN             NaN             NaN   \n",
       "87136            NaN            NaN             NaN             NaN   \n",
       "87137            NaN            NaN             NaN             NaN   \n",
       "87138            NaN            NaN             NaN             NaN   \n",
       "87139            NaN            NaN             NaN             NaN   \n",
       "\n",
       "       cf_pLDDT_90_100  \n",
       "0                  NaN  \n",
       "1                  NaN  \n",
       "2                  NaN  \n",
       "3                  NaN  \n",
       "4                  NaN  \n",
       "...                ...  \n",
       "87135              NaN  \n",
       "87136              NaN  \n",
       "87137              NaN  \n",
       "87138              NaN  \n",
       "87139              NaN  \n",
       "\n",
       "[87140 rows x 19 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_meta.isna().sum()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
